### PR TITLE
Test::Quattor: mock CAF::Path link-related functions

### LINKED ISSUE #141: Test::Quattor: mock the new symlink-related methods
Methods added in https://github.com/quattor/CAF/pull/225 should be mocked in `Test::Quattor`.

### PR DESCRIPTION
**Pull request by jouvin**
Requires https://github.com/quattor/CAF/pull/225 to pass the tests successfully.

Probably some additional unit tests would be useful but main features of link-related functions are covered.

Fixes #141

### CODE REVIEW

File: build-scripts/src/main/perl/Test/Quattor.pm [hunk 1]
```diff
@@ -520,6 +523,139 @@ Return the mocked C<is_any>
...
+$cpath->mock('_make_link', sub {
+    my ($self, $target, $link_path, %opts) = @_;
+    $link_path = sane_path($link_path);
+    $target = sane_path($target);
+
+    # Check options passed
+    Readonly my @MAKE_LINK_vALID_OPTS => qw(hard backup nocheck force);
+    for my $opt (sort keys %opts) {
+        unless ( grep (/^$opt$/, @MAKE_LINK_vALID_OPTS) ) {
+            $self->error("Invalid option ($opt) passed to _make_link()");
```
Comment (stdweird) [inline build-scripts/src/main/perl/Test/Quattor.pm]: use a `ok(0, '...')` to have something fail while running test.

File: build-scripts/src/main/perl/Test/Quattor.pm [hunk 2]
```diff
@@ -520,6 +523,139 @@ Return the mocked C<is_any>
...
+
+=cut
+
+$cpath->mock('_make_link', sub {
+    my ($self, $target, $link_path, %opts) = @_;
+    $link_path = sane_path($link_path);
+    $target = sane_path($target);
+
+    # Check options passed
+    Readonly my @MAKE_LINK_vALID_OPTS => qw(hard backup nocheck force);
```
Comment (stdweird) [inline build-scripts/src/main/perl/Test/Quattor.pm]: `vALID` -> `VALID`

File: build-scripts/src/main/perl/Test/Quattor.pm [hunk 3]
```diff
@@ -520,6 +523,139 @@ Return the mocked C<is_any>
...
+
+    my $target_full_path = $target;
+    unless ( $target_full_path =~ qr{^/} ) {
+        $target_full_path = cwd() . "/$target_full_path";
+    }
+
+    # Check that target exists if it is a hardlink or if option 'nocheck' is false
+    if ( $opts{hard} or ! $opts{nocheck} ) {
+        unless ( $files_contents{$target_full_path} ) {
+            $self->error("Symlink target ($target_full_path) doesn't exist");
```
Comment (stdweird) [inline build-scripts/src/main/perl/Test/Quattor.pm]: use `ok(0, ...)`

File: build-scripts/src/main/perl/Test/Quattor.pm [hunk 4]
```diff
@@ -520,6 +523,139 @@ Return the mocked C<is_any>
...
+    if ( $self->$is_link_method($link_path) ) {
+        if ( ($files_contents{$link_path} =~ qr/^$link_pattern(\S+)/) && ($1 eq $target) ) {
+            # Link already properly defined
+            return SUCCESS;
+        }; 
+    }
+    if ( ! $self->$is_link_method($link_path) ) {
+        if ( $self->file_exists($link_path) ) {
+            unless ( $opts{force} ) {
+                $self->error("File $link_path already exists and option 'force' not specified");
```
Comment (stdweird) [inline build-scripts/src/main/perl/Test/Quattor.pm]: `ok(0, ...)`

File: build-scripts/src/main/perl/Test/Quattor.pm [hunk 5]
```diff
@@ -520,6 +523,139 @@ Return the mocked C<is_any>
...
+                return
+            }
+        } elsif ( $self->any_exists($link_path) ) {
+            $self->error("$link_path already exists and is not a symlink");
```
Comment (stdweird) [inline build-scripts/src/main/perl/Test/Quattor.pm]: `ok(0, ..)`

**Comment by stdweird** [comment]
retest this please

**Comment by stdweird** [comment]
@jouvin jenkins test fails
```
#   Failed test 'File /some/where/files/target3 already exists and option 'force' not specified'
#   at /data/jenkins/workspace/maven-tools-pr-builder/build-scripts/src/main/perl/Test/Quattor.pm line 645.
```

**Comment by stdweird** [comment]
@jouvin wrt failing test, if you want to check that a unittest failure is triggered by the `ok(0, ...)` code, you need to mock `ok` itself. see eg https://github.com/quattor/maven-tools/blob/master/build-scripts/src/test/perl/quattor.t#L41 you can mock it before the test or method call; and unmock it again after.

**Comment by jouvin** [comment]
I'll check! It was working for me as far as I remember...

**Comment by stdweird** [comment]
big thanks!

**Pull request merged**